tests/test_node.py

```diff
@@ -74,6 +77,97 @@ def test_no_samples():
     os.chdir(cwd)
 
 
+def test_extract_clients_no_file():
+    index_ts_path = pathlib.Path(
+        FIXTURES / "node_templates" / "index_samples" / "no_exist_index.ts"
+    )
+
+    with pytest.raises(FileNotFoundError):
+        clients = node.extract_clients(index_ts_path)
+        assert not clients
+
+
+def test_extract_single_clients():
+    index_ts_path = pathlib.Path(
+        FIXTURES / "node_templates" / "index_samples" / "single_index.ts"
+    )
+
+    clients = node.extract_clients(index_ts_path)
+
+    assert len(clients) == 1
+    assert clients[0] == "TextToSpeechClient"
+
+
+def test_extract_multiple_clients():
+    index_ts_path = pathlib.Path(
+        FIXTURES / "node_templates" / "index_samples" / "multiple_index.ts"
+    )
+
+    clients = node.extract_clients(index_ts_path)
+
+    assert len(clients) == 2
+    assert clients[0] == "StreamingVideoIntelligenceServiceClient"
+    assert clients[1] == "VideoIntelligenceServiceClient"
+
+
+def test_generate_index_ts():
+    cwd = os.getcwd()
+    # use a non-nodejs template directory
+    os.chdir(FIXTURES / "node_templates" / "index_samples")
+    node.generate_index_ts(["v1", "v1beta1"], "v1")
+    generated_index_path = pathlib.Path(
+        FIXTURES / "node_templates" / "index_samples" / "src" / "index.ts"
+    )
+    sample_index_path = pathlib.Path(
+        FIXTURES / "node_templates" / "index_samples" / "sample_index.ts"
+    )
+    assert filecmp.cmp(generated_index_path, sample_index_path)
+    os.chdir(cwd)
```

Comment:
Sorry I didn't notice this sooner, but if the assertion fails on line 124, then line 125 will not execute, and tests that run later may fail because they assume they're running in a different directory.
Please add a `try:` statement before line 117, and a `finally:` statement before line 127.

Then repeat for the tests below.

Reply:
I can definitely make the changes. Just curious here. I try to break the assert `line124`, it doesn't affect the rest of test. And looks like other tests in this file and other test files (`test_samples.py`, `test_template.py`...) has same situation.

Reply:
If other tests pass after this one fails, that means we're lucky because tests don't depend on which directory they start in.

As a rule, a test should always finish with the same state as it started with.  For this test, that's true when the test passes, but not true when the assertion fails.

Reply:
I think @SurferJeffAtGoogle's suggestion here is good from a readability point of view, and I see the pattern being used in [nox](https://github.com/theacodes/nox/blob/360c0e82c8c212a13d1603b399a565856b7640aa/tests/test_tox_to_nox.py#L27) (_I think the runner we use?_).

I wonder if something has monkey patched `os.chdir` and made it return to root automatically?

Reply:
Thanks for explaining. Lesson learned 👍